crates/extricrate/src/lib.rs

```diff
@@ -384,24 +437,117 @@ pub mod dependencies {
         module_dependencies
     }
 
+    #[derive(Debug, Error)]
+    pub enum CreateCrateError {
+        #[error("failed to create crate {0}")]
+        FailedToCreateCrate(std::io::Error),
+        #[error("invalid path")]
+        InvalidPath,
+        #[error("failed to serialize Cargo.toml")]
+        InvalidCargoToml,
+    }
+
+    /// Creates a new crate named [target_crate_name] at the [target_crate_root]
+    pub fn create_target_crate(
+        target_crate_root: &std::path::Path,
+        target_crate_name: &str,
+    ) -> Result<(), CreateCrateError> {
+        create_dir_all(target_crate_root.join("src")).map_err(|_| CreateCrateError::InvalidPath)?;
+
+        let manifest = Manifest {
+            package: Some(Package::<()>::new(target_crate_name.to_owned(), "0.0.1")),
+            ..Default::default()
+        };
+
+        fs::write(
+            target_crate_root.join("Cargo.toml"),
+            toml::to_string_pretty(&manifest).map_err(|_| CreateCrateError::InvalidCargoToml)?,
+        )
+        .map_err(CreateCrateError::FailedToCreateCrate)?;
+
+        fs::write(
+            target_crate_root.join("src/lib.rs"),
+            "// your new library crate\n",
+        )
+        .map_err(CreateCrateError::FailedToCreateCrate)?;
+        Ok(())
+    }
+
+    #[derive(Error, Debug, PartialEq)]
+    pub enum GetAllModuleFilesError {
+        #[error("empty module name")]
+        EmptyModuleName,
+        #[error("failed to convert module to path {0}")]
+        ModulePathError(ListUseStatementError),
+        #[error("module is not self contained - contains non-descendants modules {0}")]
+        ModuleIsNotSelfContained(ModuleList),
+        #[error("invalid parent directory")]
+        InvalidParentDirectory,
+    }
+
+    /// Returns all the files in the create that are dependencies of `module`
+    pub fn get_all_module_files(
+        crate_root: &Path,
+        module: &ModulePath,
+        use_statements: &UseStatementMap,
+    ) -> Result<PathBuf, GetAllModuleFilesError> {
+        let mut parts = module.0.split('.').collect::<Vec<_>>();
+        let module_name = parts.pop().ok_or(GetAllModuleFilesError::EmptyModuleName)?;
```

Comment:
Is `parts` only `mut` because of this `pop`? In that case you can write it:

```suggestion
        let (parent_parts, module_name) = {
            let mut parts = module.0.split('.').collect::<Vec<_>>();
            let module_name = parts.pop().ok_or(GetAllModuleFilesError::EmptyModuleName)?;
            (parts, module_name)
        };
```

Same below with the `root_path`

Reply:
this actually made me discover `split_last` https://doc.rust-lang.org/std/primitive.slice.html#method.split_last that should work perfectly to avoid mutability